### DISCUSSION TITLE
【分享】一个只有 4KB 的 DFU bootloader

### DISCUSSION THREAD

**Discussion by xingrz** [peng-zhihui/HelloWord-Keyboard#77] [Show and tell]
[上一篇](https://github.com/peng-zhihui/HelloWord-Keyboard/discussions/76)我分享了怎么用 pyOCD 烧录程序。但如果以后更新固件还得拆开接线就很不优雅。稚晖君之前在 README 里预告了以后会做一个 bootloader，但我已经等不及了，移植了一个只有 4KB 的 DFU bootloader，分享给大家。

项目地址：https://github.com/xingrz/stm32-dfu-bootloader

### 特色

1. 只有 4KB！
2. 使用通用的 DFU 协议，随便找一个 DFU 工具就可以烧录，不限制平台

### 上手

STM32F103 空片是没有 DFU bootloader 的，所以第一次还是需要用调试器将 bootloader 烧到 STM32 里。这里我继续用 DAPLink + pyOCD。

1. 按照[上一篇分享](https://github.com/peng-zhihui/HelloWord-Keyboard/discussions/76)里介绍的步骤准备好 DAPLink 和 pyOCD 以及支持包
2. 点开我那个项目的 [Releases](https://github.com/xingrz/stm32-dfu-bootloader/releases/latest) 页面，下载 `bootloader-helloword.hex`
3. 执行下面命令刷入：

    ```
    pyocd load --pack Keil.STM32F1xx_DFP.2.4.0.pack --target stm32f103cb bootloader-helloword.hex
    ```

好，完成！之后更新固件就不需要借助调试器了。

### 如何更新固件

1. 拔掉键盘的 USB
2. 按住 Fn 键（我看原理图，应该是 Fn 键…）
3. 插上 USB，松开 Fn 键，打开 DFU 工具烧！完事

至于 DFU 工具，大家搜索一下应该有很多现成的。我自己就直接用 `dfu-util` 命令行：

```sh
dfu-util -a 0 --dfuse-address 0x08001000 -D HelloWord-Keyboard-fw.bin
```

另外，我还找到了一个可以直接在 Chrome 浏览器里用的 [webdfu](https://devanlai.github.io/webdfu/dfu-util/)，亲测也是可以用的。

<img width="812" alt="截屏2022-08-18 15 16 37" src="https://user-images.githubusercontent.com/288288/185333849-a842d10e-686d-4e18-adc0-c72b211651ec.png">

以上就是我的分享。欢迎交流。

**Comment by b1ackm4x** [comment]
这个在链接脚本里修改程序偏移的能给个示例么

**Reply by xingrz** [reply]
> 这个在链接脚本里修改程序偏移的能给个示例么

<img width="669" alt="image" src="https://user-images.githubusercontent.com/288288/202485262-9228473a-ace6-4531-aa5a-4e913010c063.png">

找一下你项目里应该会有个 .ld 格式的文件，rom 这个区段的 ORIGIN 加上 4K。

**Reply by b1ackm4x** [reply]
好的 已经成功了 多谢分享

**Comment by yr785444662** [comment]
您好，我这边bootloader和键盘固件都能成功烧录进去，但是重启后无法正确引导到键盘固件部分。能麻烦解答下或帮忙分析下吗？
情况细节：
1. 单独键盘固件可以正常使用，没有bug，大小28KB
2. 按住FN键重启电脑可以识别到DFU bootloader，也可以通过dfu-util烧写键盘固件
3. Flash是完全擦除过的，然后烧写的Bootloader和键盘固件
4. 烧写完后使用STLink Utility查看Flash内容也没有问题，从0x08000000到0x08000D30部分为BootLoader固件，从0x08001000开始是键盘固件，其余地址均为0xFF
5. 烧写的BootLoader固件是bootloader-helloword.hex这个文件，用STLink Utility烧录进去的

**Reply by xingrz** [reply]
固件本身需要在编译的时候link到正确的偏移值，并不是说你把原本在0x08000000启动的固件整体往后挪了一下就能用的。

**Reply by yr785444662** [reply]
感谢大佬回复，目前了解了一些但是不是很深入。还请大佬解答下。
请问是需要修改并编译bootloader-helloword.hex这个固件然后跳转到键盘固件的地址吗？还是键盘固件也需要修改呢？

目前我了解到的思路是先进BootLoader，如果不按FN键的话就会将程序地址跳转到键盘固件地址（0x08001000）然后开始运行键盘固件这样，所以我会以为您是将跳转地址到4K的代码写入了BootLoader固件。请问您是这样的思路吗？
